### PR TITLE
Make Densenet Scriptable

### PR DESCRIPTION
**Pull request by eellison**
A few changes: 
• Torchscript doesn't yet support checkpointing, so error if `memory_efficient` is set when trying to running a scripted model 
• Rewrite `_DenseBlock` to use `nn.Sequential`
• NYI: `any` keyword
• NYI: variadic arguments `(*args)`. Rewrote to use overloads, so that function can take in either `Tensor` or `List[Tensor]`.

### CODE REVIEW

File: torchvision/models/densenet.py [hunk 1]
```diff
@@ -66,12 +104,13 @@ def __init__(self, num_layers, num_input_features, bn_size, growth_rate, drop_ra
                 drop_rate=drop_rate,
                 memory_efficient=memory_efficient,
             )
-            self.add_module('denselayer%d' % (i + 1), layer)
+            features.append(layer)
+        self.features = nn.ModuleList(features)
```
[the remark this answers is not in the document]
Comment (eellison) [reply torchvision/models/densenet.py]: Yea help with this would be great...

**Comment by eellison** [comment]
Looks like there's a scripting error for python 2... investigating now

**Comment by eellison** [comment]
I'm on python 2 locally and can't reproduce this :(

**Comment by fmassa** [comment]
@eellison this error is very weird, I have no idea what it might be... :-/
Can you try adding some prints to the objects being compared?

**Comment by eellison** [comment]
@fmassa this is failing when python < 3 & typing isn't installed. It's hard for me to iterate on locally because `typing` is a requirement of pytorch... Since `typing` is a dependency of pytorch anyway we could add it as a dependency of torchvision too?

**Comment by fmassa** [comment]
We can install typing in the CI machines. Can you add an entry in https://github.com/pytorch/vision/blob/0dd5588207b6c0f9b3b129af7d73088e8157a153/packaging/torchvision/meta.yaml#L49 adding `typing` as well?

**Comment by eellison** [comment]
@fmassa failures now look unrelated

**Pull request merged**